core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -17,10 +17,254 @@
+        blacklistedExecsOnNode += exec
+        if (blacklistedExecsOnNode.size >= MAX_FAILED_EXEC_PER_NODE) {
+          logInfo(s"Blacklisting node $node because it has ${blacklistedExecsOnNode.size} " +
+            s"executors blacklisted: ${blacklistedExecsOnNode}")
+          nodeIdToBlacklistExpiryTime.put(node, expiryTime)
+          _nodeBlacklist.set(nodeIdToBlacklistExpiryTime.keySet.toSet)
+        }
+      }
+    }
+  }
+
+  def isExecutorBlacklisted(executorId: String): Boolean = {
+    executorIdToBlacklistStatus.contains(executorId)
+  }
+
+  /**
+   * Get the full set of nodes that are blacklisted.  Unlike other methods in this class, this *IS*
+   * thread-safe -- no lock required on a taskScheduler.
+   */
+  def nodeBlacklist(): Set[String] = {
+    _nodeBlacklist.get()
+  }
+
+  def isNodeBlacklisted(node: String): Boolean = {
+    nodeIdToBlacklistExpiryTime.contains(node)
+  }
+
+  def handleRemovedExecutor(executorId: String): Unit = {
+    // We intentionally do not clean up executors that are already blacklisted in nodeToFailedExecs,
+    // so that if another executor on the same node gets blacklisted, we can blacklist the entire
+    // node.  We also can't clean up executorIdToBlacklistStatus, so we can eventually remove
+    // the executor after the timeout.  Despite not clearing those structures here, we don't expect
+    // they will grow too big since you won't get too many executors on one node, and the timeout
+    // will clear it up periodically in any case.
+    executorIdToFailureList -= executorId
+  }
+
+
+  /**
+   * Tracks all failures for one executor (that have not passed the timeout).  Designed to
+   * efficiently remove failures that are older than the timeout, and query for the number of unique
+   * failed tasks.
+   * In general we actually expect this to be extremely small, since it won't contain more than the
+   * maximum number of task failures before an executor is failed (default 2).
+   */
+  private[scheduler] final class ExecutorFailureList extends Logging {
+
+    private case class TaskId(stage: Int, stageAttempt: Int, taskIndex: Int)
+
+    /**
+     * All failures on this executor in successful task sets, sorted by time ascending.
+     */
+    private var failures = ArrayBuffer[(TaskId, Long)]()
+
+    def addFailures(
+        stage: Int,
+        stageAttempt: Int,
+        failuresInTaskSet: ExecutorFailuresInTaskSet): Unit = {
+      // The new failures may interleave with the old ones, so rebuild the failures in sorted order.
+      // This shouldn't be expensive because if there were a lot of failures, the executor would
+      // have been blacklisted.
+      if (failuresInTaskSet.taskToFailureCountAndExpiryTime.nonEmpty) {
+        failuresInTaskSet.taskToFailureCountAndExpiryTime.foreach { case (taskIdx, (_, time)) =>
+          failures += ((TaskId(stage, stageAttempt, taskIdx), time))
```

Comment:
I don't think there is any way you would reuse a `TaskId` -- every time you're adding something here, its necessarily unique from previous additions.

In fact, maybe another question that you should be asking is, why do we even bother with `TaskId` at all?  we could instead just keep track of the failure times, and drop the ids all together -- the way this is built up, we would be certain those times correspond to unique tasks.  (In fact, you'll notice that in the existing pr, there is no de-duplication based on TaskId.)

The reason I have it is that it was useful while debugging (its referenced in the `toString`).  Even though it doesn't show up anywhere in the final code paths, I was thinking its still helpful to keep it while dealing with any future bugs that crop up.  But on second thought, maybe that is a bad idea, and we should drop it?

(`taaskToFailureCountAndExpiryTime` *does* do de-duplication, so it is important to keep at least the task index there.)